### PR TITLE
Update ErrorFormatGuesser to handle custom formats and improve guess…

### PR DESCRIPTION
**Pull request by natepage**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Wanted to implement custom error formats based on the format requested by the client, the issue was that whatever the request format the errors were always in the first error format configured. This PR has for purpose to improve the way the `ErrorFormatGuesser` is guessing the current format from the request and the configuration.

The logic is as following:
1- Check if request format has been set
2- Check if configuration for the format exist and return it if Yes
3- Otherwise guess MIME type from request headers (Accept, Content-Type)
4- Check if configuration for the MIME type exist and return it if Yes
5- Otherwise fallback to first error format configured

Example of configuration:
```yaml
api_platform:
    mapping:
        paths: ['%kernel.project_dir%/src/Database/Models']
    formats:
        json:     ['application/json']
        xml:      ['application/xml', 'text/xml']
    error_formats:
        custom_error_json: ['application/json']
        custom_error_xml:  ['application/xml', 'text/xml']
```

Return table

| Request format | Accept Header | Content-Type Header | Return
| -------------- | -------------- | --------------------- | -------
| / | / | / | `['key' => 'custom_error_xml', 'value' => ['application/xml', 'text/xml']]`
| custom_error_xml | / | / | `['key' => 'custom_error_xml', 'value' => ['application/xml', 'text/xml']]`
| / | application/json | / | `['key' => 'custom_error_json', 'value' => ['application/json']]`
| / | / | text/xml | `['key' => 'custom_error_xml', 'value' => ['application/xml', 'text/xml']]`
| custom_error_xml | application/json | text/xml | `['key' => 'custom_error_xml', 'value' => ['application/xml', 'text/xml']]`

The priority order to guess the format is: Request format > Accept Header > Content-Type Header > Fallback to first error format configured.

### CODE REVIEW

**Comment by natepage** [comment]
Travis check failed because their service was down yesterday, can we please restart the tests because it should pass. Thank you.

File: src/Util/ErrorFormatGuesser.php [hunk 1]
```diff
@@ -37,14 +37,35 @@ private function __construct()
     public static function guessErrorFormat(Request $request, array $errorFormats): array
     {
         $requestFormat = $request->getRequestFormat('');
-        if ('' !== $requestFormat && isset($errorFormats[$requestFormat])) {
+
+        if ('' !== $requestFormat && array_key_exists($requestFormat, $errorFormats)) {
```
Comment (dunglas) [inline src/Util/ErrorFormatGuesser.php]: Why this change? (`isset` is faster)

File: src/Util/ErrorFormatGuesser.php [hunk 2]
```diff
@@ -37,14 +37,35 @@ private function __construct()
...
-            return ['key' => $key, 'value' => $value];
+        $mimeType = self::getMimeType($request);
+
+        foreach ($errorFormats as $format => $mimeTypes) {
+            if (in_array($mimeType, $mimeTypes, true)) {
+                return ['key' => $format, 'value' => $mimeTypes];
+            }
+        }
+
+        return ['key' => key($errorFormats), 'value' => reset($errorFormats)];
```
Comment (dunglas) [inline src/Util/ErrorFormatGuesser.php]: Because we already loop over formats, you can just store the first format in a var, line 48:

```php
$requestMimeTypes = $request->getMimeTypes($request->getRequestFormat());
$defaultFormat = [];
foreach ($errorFormats as $format => $errorMimeTypes) {
    if ([] === $defaultFormat) {
        $defaultFormat = ['key' => $format, 'value' => $mimeTypes];
    }

    if (array_intersect($requestMimeTypes, $errorMimeTypes) {
        return ['key' => $format, 'value' => $mimeTypes];
    }
}

return $defaultFormat;
```

It's faster and safer in case no default format is defined.

File: src/Util/ErrorFormatGuesser.php [hunk 3]
```diff
@@ -37,14 +37,35 @@ private function __construct()
...
+    }
+
+    /**
+     * Get MIME type from the request.
+     *
+     * @param \Symfony\Component\HttpFoundation\Request $request
```
Comment (dunglas) [inline src/Util/ErrorFormatGuesser.php]: `@param` and `@return` can be removed because of the typehints.

File: src/Util/ErrorFormatGuesser.php [hunk 4]
```diff
@@ -37,14 +37,35 @@ private function __construct()
...
+     *
+     * @return string|null
+     */
+    private static function getMimeType(Request $request): ?string
```
Comment (dunglas) [inline src/Util/ErrorFormatGuesser.php]: You can remove this method and just call `$request->getMimeTypes($request->getRequestFormat())` because the content negotiation is already properly handled here: https://github.com/api-platform/core/blob/master/src/EventListener/AddFormatListener.php#L67

File: src/Util/ErrorFormatGuesser.php [hunk 5]
```diff
@@ -37,14 +37,35 @@ private function __construct()
...
+        if ('' !== $requestFormat && array_key_exists($requestFormat, $errorFormats)) {
             return ['key' => $requestFormat, 'value' => $errorFormats[$requestFormat]];
         }
 
-        foreach ($errorFormats as $key => $value) {
-            return ['key' => $key, 'value' => $value];
+        $mimeType = self::getMimeType($request);
+
+        foreach ($errorFormats as $format => $mimeTypes) {
+            if (in_array($mimeType, $mimeTypes, true)) {
```
Comment (dunglas) [inline src/Util/ErrorFormatGuesser.php]: Sometimes they are several mime types associated with one formats (it's why you should call $request->getMimeType**s**()). They all must be checked. See my alternative proposal below.

**Comment by dunglas** [comment]
Can you also target the master? It's a new feature to me.

**Comment by natepage** [comment]
@dunglas I've made the changes, thank you for the pieces of code you gave me. I don't know how to target master instead of 2.1, do you want me to create a new PR to master?

**Comment by dunglas** [comment]
@natepage `git rebase master` should do the trick.

**Comment by meyerbaptiste** [comment]
Rebased and RFR @api-platform/core-team.

File: src/Util/ErrorFormatGuesser.php [hunk 6]
```diff
@@ -37,14 +37,24 @@ private function __construct()
...
             return ['key' => $requestFormat, 'value' => $errorFormats[$requestFormat]];
         }
 
-        foreach ($errorFormats as $key => $value) {
-            return ['key' => $key, 'value' => $value];
+        $requestMimeTypes = Request::getMimeTypes($request->getRequestFormat());
+        $defaultFormat = [];
+
+        foreach ($errorFormats as $format => $errorMimeTypes) {
+            if ([] === $defaultFormat) {
```
Comment (soyuka) [inline src/Util/ErrorFormatGuesser.php]: what? this looks weird can we change to empty? or `!$defaultFormat`? Why can't we init above with the proper array?

Comment (dunglas) [reply src/Util/ErrorFormatGuesser.php]: Because we not sure the key 0 exist, it looks ok to me to do this. We indeed may use `!$defaultFormat`.

Comment (soyuka) [reply src/Util/ErrorFormatGuesser.php]: I'd be :+1: for `!$defaultFormat`. Anyway I panicked because `[] === []` is false in javascript :laughing:

Comment (meyerbaptiste) [reply src/Util/ErrorFormatGuesser.php]: Fixed.

**Comment by dunglas** [comment]
Thanks @natepage and @meyerbaptiste!
